File: deck/seed/code/native/swift/file/runtime/watch.swift
// Filesystem watching for the swift target, over a DispatchSource file-system-object source: the kqueue the
// system already runs, not a polling loop and not a package.
//
// The API is PULL, not callback: `watchOpen` starts the watch, `watchNext` awaits the next change, `watchClose`
// stops it. That is the one shape all four backends can hold (node's fs.watch async iterator, notify's channel,
// this, the JVM's WatchService), so the Term API above it is the same everywhere.
//
// NAMED GAP: kqueue reports that a watched path changed, not what changed inside it, and it cannot watch a tree.
// So `deep` is accepted and does nothing here, and a change under a watched directory arrives as one "change" on
// the directory. FSEvents is the darwin API that does both and is not portable to swift on Linux; a watcher that
// covers `deep` properly is its own piece of work, recorded in note/term/stdlib/native-async-file-and-server.md.
//
// Reached only through the public file/watch API.
import Dispatch
import Foundation

enum watch {
  // the running watch: the descriptor and its dispatch source, plus a queue of changes the source has posted and
  // nobody has taken yet, and whoever is waiting for one
  final class Watcher: @unchecked Sendable {
    let lock = NSLock()
    var descriptor: Int32 = -1
    var source: DispatchSourceFileSystemObject?
    var ready: [WatchEvent] = []
    var waiting: CheckedContinuation<WatchEvent, Never>?
    var closed = false
    let path: String

    init(path: String) {
      self.path = path
    }

    // hand the change to whoever is waiting, or hold it until someone asks
    func post(_ event: WatchEvent) {
      lock.lock()
      let waiter = waiting
      waiting = nil

      if waiter == nil {
        ready.append(event)
      }

      lock.unlock()
      waiter?.resume(returning: event)
    }
  }

  static func watchOpen(_ path: String, _ deep: Bool) async -> Watcher {
    let watcher = Watcher(path: path)
    let descriptor = open(path, O_EVTONLY)

    guard descriptor >= 0 else {
      return watcher
    }

    let source = DispatchSource.makeFileSystemObjectSource(
      fileDescriptor: descriptor,
      eventMask: [.write, .delete, .rename, .extend, .attrib],
      queue: DispatchQueue.global()
    )

    source.setEventHandler {
      let mask = source.data
      let kind: String

      if mask.contains(.delete) {
        kind = "remove"
      } else if mask.contains(.rename) {
        kind = "create"
      } else if mask.contains(.write) || mask.contains(.extend)
        || mask.contains(.attrib)
      {
        kind = "change"
      } else {
        kind = "other"
      }

      watcher.post(WatchEvent(kind: kind, path: path))
    }

    source.setCancelHandler {
      close(descriptor)
    }

    watcher.descriptor = descriptor
    watcher.source = source
    source.resume()

    return watcher
  }

  // the next change. A closed watcher answers with the empty event rather than waiting forever, so a loop over
  // `watchNext` ends after `watchClose`.
  static func watchNext(_ watcher: Watcher) async -> WatchEvent {
    await withCheckedContinuation { (hold: CheckedContinuation<WatchEvent, Never>) in
      watcher.lock.lock()

      if !watcher.ready.isEmpty {
        let event = watcher.ready.removeFirst()
        watcher.lock.unlock()
        hold.resume(returning: event)

        return
      }

      if watcher.closed {
        watcher.lock.unlock()
        hold.resume(returning: WatchEvent(kind: "", path: ""))

        return
      }

      watcher.waiting = hold
      watcher.lock.unlock()
    }
  }

  static func watchClose(_ watcher: Watcher) async {
    watcher.lock.lock()
    watcher.closed = true
    let waiter = watcher.waiting
    watcher.waiting = nil
    let source = watcher.source
    watcher.source = nil
    watcher.lock.unlock()

    source?.cancel()
    waiter?.resume(returning: WatchEvent(kind: "", path: ""))
  }
}

File: deck/seed/code/native/node/file/runtime/watch.ts
// Filesystem watching for node, over `fs.watch`, which is libuv's FSEvents / inotify binding and not a polling
// loop.
//
// The API is PULL, not callback: `watchOpen` starts the watch, `watchNext` awaits the next change, `watchClose`
// stops it. node is the ONE backend where a callback shape would have been natural, and it does not get one: the
// other three cannot hold it (a Term task value would have to cross into a platform thread), so the shape all
// four can hold is the shape all four get. What node contributes here is a queue in front of an event emitter.
//
// Reached only through the public file/watch API.
import * as fs from 'node:fs'
import * as path from 'node:path'

type Event = { kind: string; path: string }

type Watcher = {
  watcher: fs.FSWatcher | undefined
  ready: Event[]
  waiting: ((event: Event) => void) | undefined
  closed: boolean
}

const EMPTY: Event = { kind: '', path: '' }

const watch = {
  watchOpen: async (at: string, deep: boolean): Promise<Watcher> => {
    const held: Watcher = {
      watcher: undefined,
      ready: [],
      waiting: undefined,
      closed: false,
    }

    // hand the change to whoever is waiting, or hold it until someone asks
    const post = (event: Event): void => {
      const waiter = held.waiting

      if (waiter) {
        held.waiting = undefined
        waiter(event)

        return
      }

      held.ready.push(event)
    }

    try {
      held.watcher = fs.watch(at, { recursive: deep }, (event, name) => {
        // node reports 'rename' for both a create and a remove, so which one it was is whether the path is
        // still there. Anything else is a content change.
        const whole = name ? path.join(at, name.toString()) : at

        if (event === 'rename') {
          post({
            kind: fs.existsSync(whole) ? 'create' : 'remove',
            path: whole,
          })

          return
        }

        post({ kind: 'change', path: whole })
      })
    } catch {
      held.closed = true
    }

    return held
  },

  // the next change. A closed watcher answers with the empty event rather than waiting forever, so a loop over
  // `watchNext` ends after `watchClose`.
  watchNext: async (held: Watcher): Promise<Event> => {
    const already = held.ready.shift()

    if (already) {
      return already
    }

    if (held.closed) {
      return EMPTY
    }

    return new Promise<Event>(answer => {
      held.waiting = answer
    })
  },

  watchClose: async (held: Watcher): Promise<void> => {
    held.closed = true
    held.watcher?.close()
    held.watcher = undefined

    const waiter = held.waiting
    held.waiting = undefined
    waiter?.(EMPTY)
  },
}
